FWCore/Catalog/src/InputFileCatalog.cc

```diff
@@ -14,46 +17,48 @@ namespace edm {
                                      bool useLFNasPFNifLFNnotFound)
-      : logicalFileNames_(fileNames),
-        fileNames_(fileNames),
-        fallbackFileNames_(fileNames.size()),
-        fileCatalogItems_(),
-        fileLocator_(),
-        overrideFileLocator_(),
-        fallbackFileLocator_(),
-        overrideFallbackFileLocator_() {
-    init(override, "", useLFNasPFNifLFNnotFound);
-  }
-
-  InputFileCatalog::InputFileCatalog(std::vector<std::string> const& fileNames,
-                                     std::string const& override,
-                                     std::string const& overrideFallback,
-                                     bool useLFNasPFNifLFNnotFound)
-      : logicalFileNames_(fileNames),
-        fileNames_(fileNames),
-        fallbackFileNames_(fileNames.size()),
-        fileCatalogItems_(),
-        fileLocator_(),
-        overrideFileLocator_(),
-        fallbackFileLocator_(),
-        overrideFallbackFileLocator_() {
-    init(override, overrideFallback, useLFNasPFNifLFNnotFound);
+      : logicalFileNames_(fileNames), fileNames_(fileNames), fileCatalogItems_(), overrideFileLocator_() {
+    init(override, useLFNasPFNifLFNnotFound);
   }
 
   InputFileCatalog::~InputFileCatalog() {}
 
-  void InputFileCatalog::init(std::string const& inputOverride,
-                              std::string const& inputOverrideFallback,
-                              bool useLFNasPFNifLFNnotFound) {
-    fileCatalogItems_.reserve(fileNames_.size());
+  std::vector<std::string> InputFileCatalog::fileNames(unsigned iCatalog) const {
+    std::vector<std::string> tmp;
+    for (auto it = fileCatalogItems_.begin(); it != fileCatalogItems_.end(); ++it) {
+      tmp.push_back(it->fileName(iCatalog));
+    }
+    return tmp;
+  }
+
+  void InputFileCatalog::init(std::string const& inputOverride, bool useLFNasPFNifLFNnotFound) {
     typedef std::vector<std::string>::iterator iter;
-    for (iter it = fileNames_.begin(),
-              lt = logicalFileNames_.begin(),
-              itEnd = fileNames_.end(),
-              ft = fallbackFileNames_.begin();
-         it != itEnd;
-         ++it, ++lt, ++ft) {
+
+    if (!overrideFileLocator_ && !inputOverride.empty()) {
+      overrideFileLocator_ =
+          std::make_unique<FileLocator>(inputOverride);  // propagate_const<T> has no reset() function
+    }
+
+    Service<SiteLocalConfig> localconfservice;
+    if (!localconfservice.isAvailable())
+      throw cms::Exception("TrivialFileCatalog", "edm::SiteLocalConfigService is not available");
+
+    std::vector<std::string> const& tmp_dataCatalogs = localconfservice->dataCatalogs();
+    if (!fileLocators_.empty())
+      fileLocators_.clear();
+
+    for (auto it = tmp_dataCatalogs.begin(); it != tmp_dataCatalogs.end(); ++it) {
+      try {
+        fileLocators_.push_back(std::make_unique<FileLocator>(*it));
+      } catch (cms::Exception const& e) {
+        continue;
+      }
```

Comment:
It is a good point. I will throw exception if none of data catalogs work

Reply:
Please see this
FWCore/Catalog/src/InputFileCatalog.cc
for new implementation

Reply:
Looking at the constructor of `FileLocator`, ignoring the exceptions from it could easily hide e.g. typos in the site local config xml. For example, having typos in all but last catalog entries would go unnoticed leading to everything to use the last catalog (hopefully catched from monitoring, but I would imagine earlier notification could be helpful).

Closest for preserving the current behavior would be to require the first `FileLocator` constructor to succeed, and silently ignore the failures of later `FileLocator` constructors.

But I'm not really sure what would be the best option.

Reply:
It is a good point! I think it makes absolutely sense to enforce the first FileLocator constructor to succeed. Although we provide a framework to allow flexibility in handling the file reading/opening, users (who design site local config xml) should bear in mind that they should make their best bet on the first data catalog for it to work. Naive mistakes like typos should be notified right away as you pointed out. We can print warnings if other data catalogs fail. If you agree with my suggestion, please let me know.

Reply:
Hi @makortel, I changed the InputFileCatalog::init to require the first FileLocator constructor to succeed and silently ignore other FileLocator constructor failures if the corresponding dataCatalogs are broken as you have suggested.